### DISCUSSION TITLE
Overlap library bind for python 2.7

### DISCUSSION THREAD

**Discussion by alisheikholeslam** [severinstrobl/overlap#97] [General]
Hello dear Severin Strobl,

I have used your prepared overlap library for Python Ver. 3+ several times; Now I am using a software that embeds Python Ver. 2.7 and I must adjust the library for this Python version. Have you prepared any Python 2.7 version compatible one? I am not familiar with and do not have any experience with binding and ... to do it myself now and it will be very helpful if the compatible version exists.

Best regards
Ali

**Comment by severinstrobl** [comment]
Hi @alisheikholeslam, sorry for my delayed response, I was unavailable due to travelling. Unfortunately Python 2.7 is not supported out of the box, as [pybind11](https://github.com/pybind/pybind11), the library used to create the Python bindings for the overlap library, does not support Python 2.7 any more. The last version to support Python 2.7 is [v2.9.2](https://github.com/pybind/pybind11/releases/tag/v2.9.2). It should be possible to use this older version of pybind11 to build the bindings, but some (minor?) tweaks might be required. What platform are you on, Linux x86-64?

**Reply by alisheikholeslam** [reply]
@severinstrobl Thank you for your response and the note, I am working on windows 10 x64. Do you have any plan to release a Python 2.7 compatible version of that?

**Comment by severinstrobl** [comment]
On Linux x86-64 it was easy to build the library for Python 2.7 with minor tweaks to the project setup, but I did not try Windows 10 yet. I'm not sure about releasing prebuild wheels for Python 2.7, after all Python 2.7 reached its end-of-life more than 4 years ago... I'll see how easy it is to build the library for 2.7 on all platforms in an automated way and decide based on that.

**Reply by severinstrobl** [reply]
Hi @alisheikholeslam, took me a while to dig into this in more detail, sorry for the delay. After spending a significant amount of time trying to figure out a way to build the wheels for Python 2.7 on Windows, here are my findings:

1. Compiling Python 2.7 extensions on Windows requires a very specific version of the MS VS C++ compiler ([v9.0](https://wiki.python.org/moin/WindowsCompilers#Microsoft_Visual_C.2B-.2B-_9.0_standalone:_Visual_C.2B-.2B-_Compiler_for_Python_2.7_.28x86.2C_x64.29)).
2. This compiler is simply not around any more. There are some archived copies here and there, but nothing official.
3. Extending the automated build pipeline to include Python 2.7 on Windows is thus simply not possible.

So, unfortunately, this means there will be no wheels from my side for Python 2.7 on Windows (Linux/MacOS seem to be less problematic).
This leaves you with 3 options:

1. You try to build the library on your own system for Python 2.7 on Windows (should in principle not be too hard, assuming you still find a working copy of the old MS compiler).
2. You switch to a different OS (Linux/MacOS) where compiling extensions for Python 2.7 still works.
3. (Preferable!) You switch to Python 3. Python 2.7 has been deprecated for many years, if possible at all, you should update your software stack.

**Reply by alisheikholeslam** [reply]
Thanks a lot, @severinstrobl, for your follow-up. I did it in Ubuntu and I am using it now. It was a little challenging for me due to my little experience working with Ubunto and ... . I have tried [**VCForPython27**](https://web.archive.org/web/20210106040224/https://download.microsoft.com/download/7/9/6/796EF2E4-801B-4FC4-AB28-B59FBF6D907B/VCForPython27.msi), but it doesn't work. I guess other tools than that must be used e.g. **minGW**, **swing**, or ... .

Yours sincerely
